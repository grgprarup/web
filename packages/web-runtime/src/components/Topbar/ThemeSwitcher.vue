<template>
  <oc-button
    v-oc-tooltip="buttonLabel"
    class="themeswitcher-btn"
    :aria-label="buttonLabel"
    appearance="raw"
    variation="inverse"
    @click="changeTheme"
  >
    <span class="oc-visible@s" :aria-label="switchLabel" v-text="switchText" />
    <oc-icon :name="switchIcon" fill-type="line" />
  </oc-button>
</template>
<script>
import { unref, watch } from '@vue/composition-api'
import { mapGetters } from 'vuex'
import { useStore, useLocalStorage } from 'web-pkg/src/composables'

export default {
  computed: {
    ...mapGetters(['configuration']),
    buttonLabel() {
      return this.$gettext('Click to switch theme')
    },
    switchIcon() {
      return this.mode === 'light' ? 'sun' : 'moon-clear'
    },
    switchLabel() {
      return this.$gettext('Currently used theme')
    },
    switchText() {
      return this.mode === 'light' ? this.$gettext('Light mode') : this.$gettext('Dark mode')
    }
  },
  setup() {
    const store = useStore()
    const mode = useLocalStorage('oc_colorMode', null)
    const currentTheme = useLocalStorage(
      'oc_currentTheme',
      store.getters.configuration.currentTheme
    )

    watch(
      currentTheme,
      () => {
        store.dispatch('loadTheme', { theme: unref(currentTheme) })
      },
      { immediate: true }
    )

    return { currentTheme, mode }
  },
  mounted() {
    if (this.mode === null) {
      this.mode = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light'
      const themeToLoad = window.matchMedia('(prefers-color-scheme: dark)').matches
        ? 'default-dark'
        : 'default'
      this.applyTheme(themeToLoad, this.mode)
    }
  },
  methods: {
    changeTheme() {
      if (this.mode === 'light' || this.mode === null) {
        this.applyTheme('default-dark', 'dark')
      } else {
        this.applyTheme('default', 'light')
      }
    },

    applyTheme(themeName, themeMode) {
      this.currentTheme = this.configuration.themes[themeName]
      this.mode = themeMode
      this.applyChangedTheme()
    },

    applyChangedTheme() {
      for (const param in this.currentTheme.designTokens.colorPalette) {
        document
          .querySelector(':root')
          .style.setProperty(
            '--oc-' + 'color-' + param,
            this.currentTheme.designTokens.colorPalette[param]
          )
      }
    }
  }
}
</script>
