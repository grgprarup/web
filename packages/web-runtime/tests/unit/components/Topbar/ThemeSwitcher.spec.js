import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import GetTextPlugin from 'vue-gettext'
import DesignSystem from 'owncloud-design-system'

import ThemeSwitcher from 'web-runtime/src/components/Topbar/ThemeSwitcher.vue'
import stubs from '../../../../../../tests/unit/stubs'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const lightTheme = {
  designTokens: {
    colorPalette: {
      'background-accentuate': 'lime'
    }
  }
}

const darkTheme = {
  designTokens: {
    colorPalette: {
      'background-accentuate': 'gold'
    }
  }
}

const spyChangeTheme = jest.spyOn(ThemeSwitcher.methods, 'changeTheme')

describe('ThemeSwitcher component', () => {
  describe('visually', () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('renders a button, initially in light mode per default', async () => {
      window.matchMedia = darkModePreferred(false)
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })

    it('renders a button, initially in dark mode if user prefers dark color sceme', async () => {
      window.matchMedia = darkModePreferred(true)
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
      // await expect(spyChangeTheme).toHaveBeenCalledTimes(1)
    })

    it('toggles between themes upon click', async () => {
      window.matchMedia = darkModePreferred(false)
      const wrapper = getWrapper()
      expect(spyChangeTheme).toHaveBeenCalledTimes(0)

      await wrapper.find('.themeswitcher-btn').trigger('click')
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
      expect(spyChangeTheme).toHaveBeenCalledTimes(1)

      await wrapper.find('.themeswitcher-btn').trigger('click')
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
      expect(spyChangeTheme).toHaveBeenCalledTimes(2)
    })
  })

  describe('restores', () => {
    it('light theme if light theme is saved in localstorage', async () => {
      window.localStorage.setItem('oc_currentTheme', JSON.stringify(lightTheme))
      window.localStorage.setItem('oc_colorMode', 'light')
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })
    it('dark theme if dark theme is saved in localstorage', async () => {
      window.localStorage.setItem('oc_currentTheme', JSON.stringify(darkTheme))
      window.localStorage.setItem('oc_colorMode', 'dark')
      const wrapper = getWrapper()
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })
  })
})

function darkModePreferred(enabled = false) {
  return function () {
    return {
      matches: enabled,
      media: "'(prefers-color-scheme: dark)'",
      addListener: function () {},
      removeListener: function () {}
    }
  }
}

function getWrapper() {
  return mount(ThemeSwitcher, {
    store: new Vuex.Store({
      getters: {
        configuration: () => ({
          themes: {
            default: lightTheme,
            'default-dark': darkTheme
          },
          currentTheme: {
            designTokens: {
              colorPalette: {
                'background-accentuate': 'lime'
              }
            }
          }
        })
      },
      actions: {
        loadTheme: jest.fn()
      }
    }),
    localVue,
    stubs: { ...stubs, 'oc-button': false },
    directives: {
      'oc-tooltip': jest.fn()
    }
  })
}
