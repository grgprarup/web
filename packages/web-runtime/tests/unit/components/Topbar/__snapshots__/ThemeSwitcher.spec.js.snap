// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher component restores dark theme if dark theme is saved in localstorage 1`] = `
<button aria-label="Click to switch theme" class="themeswitcher-btn oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Dark mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component restores light theme if light theme is saved in localstorage 1`] = `
<button aria-label="Click to switch theme" class="themeswitcher-btn oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Light mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually renders a button, initially in dark mode if user prefers dark color sceme 1`] = `
<button aria-label="Click to switch theme" class="themeswitcher-btn oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Dark mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually renders a button, initially in light mode per default 1`] = `
<button aria-label="Click to switch theme" class="themeswitcher-btn oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Light mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually toggles between themes upon click 1`] = `
<button aria-label="Click to switch theme" class="themeswitcher-btn oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Dark mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub>
</button>
`;

exports[`ThemeSwitcher component visually toggles between themes upon click 2`] = `
<button aria-label="Click to switch theme" class="themeswitcher-btn oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw"><span aria-label="Currently used theme" class="oc-visible@s">Light mode</span>
  <oc-icon-stub name="moon-clear" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive"></oc-icon-stub>
</button>
`;
